### PR TITLE
Use @pixi-build-tools/api-extractor-lerna-monorepo for typings!

### PR DESCRIPTION
**Pull request by SukantPal**
# Brief 🎢 

~~Blocked by #6848~~
Blocked by #6857 

This will use [@pixi-build-tools/api-extractor-lerna-monorepo](https://www.npmjs.com/package/@pixi-build-tools/api-extractor-lerna-monorepo) to generate typings in our monorepo using [@microsoft/api-extractor](https://www.npmjs.com/package/@microsoft/api-extractor).

## Caveats 😢 

* ~~`pixi.js` and `pixi.js-legacy` fail because "Unsupported export defaultVertex"~~

```
Analysis will use the bundled TypeScript version 3.9.5
pixi.js-legacy failed:
Error: Unsupported export "defaultVertex":
/Users/shukantpal/Web Projects/pixi.js/global.d.ts:8:11
    at ExportAnalyzer.fetchAstModuleFromSourceFile (/Users/shukantpal/Web Projects/pixi.js/node_modules/@microsoft/api-extractor/lib/analyzer/ExportAnalyzer.js:74:35)
    at ExportAnalyzer._fetchSpecifierAstModule (/Users/shukantpal/Web Projects/pixi.js/node_modules/@microsoft/api-extractor/lib/analyzer/ExportAnalyzer.js:569:41)
    at ExportAnalyzer.fetchAstModuleFromSourceFile (/Users/shukantpal/Web Projects/pixi.js/node_modules/@microsoft/api-extractor/lib/analyzer/ExportAnalyzer.js:90:65)
    at AstSymbolTable.fetchAstModuleFromWorkingPackage (/Users/shukantpal/Web Projects/pixi.js/node_modules/@microsoft/api-extractor/lib/analyzer/AstSymbolTable.js:57:37)
    at Collector.analyze (/Users/shukantpal/Web Projects/pixi.js/node_modules/@microsoft/api-extractor/lib/collector/Collector.js:117:51)
    at Function.invoke (/Users/shukantpal/Web Projects/pixi.js/node_modules/@microsoft/api-extractor/lib/api/Extractor.js:102:19)
    at /Users/shukantpal/Web Projects/pixi.js/node_modules/@pixi-build-tools/api-extractor-lerna-monorepo/index.js:92:37
    at Array.forEach (<anonymous>)
    at main (/Users/shukantpal/Web Projects/pixi.js/node_modules/@pixi-build-tools/api-extractor-lerna-monorepo/index.js:41:16)
```

I have no idea how to fix that problem, I'll probably consult with the api-extractor team.

* ~~`@pixi/core`, `@pixi/graphics`: "Error: "import * as ___ from ___;" is not supported yet for local files."~~

We can probably work around this by importing everything by name, like:

```js
import * as resources from './texture/resources';
```

⏬ 

```js
import { BufferResource, BlobResource, ... } from './texture/resources';

const resources = { BufferResource, BlobResource, ... };
```

##### Pre-Merge Checklist
- [ ] Lint process passed (`npm run lint`)

### CODE REVIEW

**Comment by ivanpopelyshev** [comment]
Agree, we can change that `resources` place

File: package.json [hunk 1]
```diff
@@ -14,27 +14,21 @@
...
-    "types:legacy": "mkdirp dist/types && jsdoc -c types/jsdoc-legacy.conf.json && node types/assemble pixi.js-legacy",
-    "posttypes": "rimraf dist/compiled",
     "lint": "eslint --ext .js --ext .ts test bundles packages tools --ignore-path .gitignore --max-warnings 0",
     "lintfix": "npm run lint -- --fix",
     "prebuild": "npm run clean:build",
     "build": "npm run build:dev",
     "build:prod": "cross-env NODE_ENV=production rollup -c",
     "build:dev": "rollup -c",
-    "postbuild": "npm run types",
+    "build:types": "rm -r out; tsc; api-extractor-lerna-monorepo; rm -r out",
```
Comment (Zyie) [inline package.json:23]: This doesn't seem to work for me on windows as `rm` is not a command

Can this be replaced with using `rimraf`
 `"build:types": "rimraf out && tsc && api-extractor-lerna-monorepo && rimraf out",`

**Comment by Zyie** [comment]
So not sure if this is the same thing but in other type generation PR they added this code [snippet](https://github.com/pixijs/pixi.js/pull/6484/files#diff-0706e21a0b4b7fcff33a8f4e5f2b3aafR53-R63) to work around a problem of `api-extractor` not following the declares of `*.vert` and `*.frag` in `global.d.ts` 

Maybe this is the cause of your `defaultVertex` issues?

**Comment by bigtimebuddy** [comment]
Can you rebase to *dev*?

**Comment by SukantPal** [comment]
@bigtimebuddy @Zyie I found hacky workarounds.

## Shader files

You add an indirection to the import, and explicitly type them as a `string`:

https://github.com/pixijs/pixi.js/blob/e4da078e2395d084b50d7f32d1143cd74e7e7a9a/packages/core/src/fragments/index.ts#L18

## systems, resources namespace

See, api-extractor doesn't support `import * as __ from ___;` right, so you have to import all of the classes by name. The problem is if you export them as a `const`, api-extractor can't output a `namespace`. That means you can't use `systems.System` as a type (only as a value). The hack for this is to output a `const` and an `interface`. You can then use `systems['FilterSystem']` as a type (not `systems.FilterSystem` sadly):

```ts
export const systems = {
    FilterSystem,
    BatchSystem,
    ContextSystem,
    FramebufferSystem,
    GeometrySystem,
    MaskSystem,
    ScissorSystem,
    StencilSystem,
    ProjectionSystem,
    RenderTextureSystem,
    ShaderSystem,
    StateSystem,
    TextureGCSystem,
    TextureSystem
};

export interface systems {
    FilterSystem: FilterSystem;
    BatchSystem: BatchSystem;
    ContextSystem: ContextSystem;
    FramebufferSystem: FramebufferSystem;
    GeometrySystem: GeometrySystem;
    MaskSystem: MaskSystem;
    ScissorSystem: ScissorSystem;
    StencilSystem: StencilSystem;
    ProjectionSystem: ProjectionSystem;
    RenderTextureSystem: RenderTextureSystem;
    ShaderSystem: ShaderSystem;
    StateSystem: StateSystem;
    TextureGCSystem: TextureGCSystem;
    TextureSystem: TextureSystem;
}
```

Usage:

```ts
import { systems } from '@pixi/core';

function (param: systems['FilterSystem']) {
 const value = new systems.FilterSystem();
}
```

@bigtimebuddy Is this a reasonable tradeoff?

**Comment by SukantPal** [comment]
@ivanpopelyshev Please review. I'll remove the .d.ts files once we decide they are good.

**Comment by bigtimebuddy** [comment]
Alternatively, we could potentially _remove_ `systems` and `resources` namespaces. I've been slowly trying to kill the internal namespaces like interaction, accessibility, extract. Maybe systems and resources aren't that meaningful organizationally? Thoughts?

**Comment by SukantPal** [comment]
@bigtimebuddy I think you could potentially do that after this PR. It would be a bit messy if I put it here.

**Comment by ivanpopelyshev** [comment]
@bigtimebuddy  +1!

I want to move out Filters & Masks away at some point, so yeah, internal namespaces there are not needed. 

Also, ArrayResource and that kind of resources have to go away.

**Comment by bigtimebuddy** [comment]
@SukantPal do you mind updating to the latest and resolving these conflicts?

**Comment by SukantPal** [comment]
Will do!

On Sep 4, 2020, at 9:52 AM, Matt Karl <notifications@github.com<mailto:notifications@github.com>> wrote:



@SukantPal<https://github.com/SukantPal> do you mind updating to the latest and resolving these conflicts?

**Comment by SukantPal** [comment]
@bigtimebuddy @ivanpopelyshev I removed the declaration files from version control (and stripped usage of systems, resources namespaces). This should be good to go!

**Review by bigtimebuddy — APPROVED** [review]

**Comment by ivanpopelyshev** [comment]
Gonna check it tomorrow

**Comment by SukantPal** [comment]
As a side note, I think, in the future, we should remove the `PIXI` namespace from the documentation as well.

**Comment by bigtimebuddy** [comment]
@SukantPal I agree moving away from the PIXI global is better. We, however, need a way to denote the module a class is from. Does webdoc have something like that?

**Comment by SukantPal** [comment]
Glad you asked! I and @Zekfad are currently working on getting a 1.1.0 release done for webdoc (I'm aiming by this month).

pixi-webdoc-template would most likely only get a note saying "package: @pixi/core" (or whatever package). (you can get the package name via `doc.loc.file.package.metadata.name`, an unreleased API)

The default-template I'm working has a hierarchical api-explorerer menu similar to [this](https://rushstack.io/pages/api/rush-lib/). I can probably show you how it would look like next week.

The following PR would introduce support for multi-package mode:

https://github.com/webdoc-js/webdoc/pull/32

**Comment by bigtimebuddy** [comment]
I'm all for this, thanks for the update.

File: package.json [hunk 2]
```diff
@@ -20,10 +20,11 @@
     "build": "npm run build:dev",
     "build:prod": "cross-env NODE_ENV=production rollup -c",
     "build:dev": "rollup -c",
+    "build:types": "rimraf out; tsc; api-extractor-lerna-monorepo; rimraf out",
```
Comment (Zyie) [inline package.json:23]: Just did a fresh install of the repo and this still doesn't work on windows. I don't think `;` separated commands work. They need to be `&&`

**Review by Zyie — CHANGES_REQUESTED** [review]
This branch isn't building on windows. I left a comment to fix the npm command. However i also get the following error when running the updated command

`node_modules/resource-loader/typings/resource-loader.d.ts:505:24 - error TS2304: Cannot find name 'Signal'.`

This can be fixed by adding the `skipLibCheck` in the `tsconfig`.

Other comments:

- You need to update the `types` and `files` in the `package.json` for both `pixi.js` and `pixi.js-legacy` since the file name is now `index.d.ts` instead of `pixi.js.d.ts`.

- Are we not going to be releasing the types per package? e.g. if i download just the essential packages from pixi-customize i wouldn't get any types since they are only in the `pixi.js` repo?

- These types dont seem to include any of the `GlobalMixins` stuff. For example the `canvas-display` types file is empty when its should declare the changes it makes to `Container`

**Comment by ivanpopelyshev** [comment]
windows, ts 3.8.3 on pc, 3.9.5 in node_modules.

```js
node_modules/mini-signals/typings/mini-signals.d.ts:19:2 - error TS2309: An export assignment cannot be used in a module with other exported elements.
```

File: package.json (hunk 2 above)
Comment (SukantPal) [reply package.json:23]: Done!

**Comment by SukantPal** [comment]
@Zyie 

The typings are to be shipped on a per-package basis. This PR does not address the `GlobalMixins` problem. I followed up on all the other issues.

**Review by Zyie — APPROVED** [review]

**Comment by SukantPal** [comment]
@ivanpopelyshev Try again!!!

**Comment by ivanpopelyshev** [comment]
Its slow as heck. We'll have to profile it later.

How are we supposed to test this titanic work properly:

1. Find a way to npm link everything, set up a ts project that uses it
2. Check that all `global.d.ts` gets referenced. somehow.
3. Generate typings for `pixi.js` and `pixi-legacy.js`, npm link and test them in ts project
4. test it with vanilla typescript, no es6 modules.

**Comment by ivanpopelyshev** [comment]
All of that would be done in separate PR's but here we have a problem: once we merge this thing, our usual typings will stop working because `build:types` task was changed.

**Comment by bigtimebuddy** [comment]
FYI, our old typings have already been removed. This PR re-adds the types.

**Comment by SukantPal** [comment]
@bigtimebuddy This should be "ready to merge"?

**Review by ivanpopelyshev — APPROVED** [review]
merge first, test later

**Comment by bigtimebuddy** [comment]
I have outstanding questions about this PR that I'd like to address in follow-up:
* Fix the global mixins and make sure they work
* Should we deploy types to CDN

**Pull request merged**